joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForFunctionsCreator.scala

```diff
@@ -2,64 +2,116 @@ package io.joern.gosrc2cpg.astcreation
+    node: ParserNodeInfo,
+    method: NewMethod,
+    methodName: String,
+    methodFullName: String,
+    signature: String
+  ): Ast = {
+
+    val parentNode: NewTypeDecl = methodAstParentStack.collectFirst { case t: NewTypeDecl => t }.getOrElse {
+      // TODO: Need to add respective Unit test to test this possibility, as looks to me as dead code. Replicated it from 'c2cpg' by referring AstForFunctionsCreator.
+      val astParentType     = methodAstParentStack.head.label
+      val astParentFullName = methodAstParentStack.head.properties("FULL_NAME").toString
+      val typeDeclNode_ =
+        typeDeclNode(node, methodName, methodFullName, method.filename, methodName, astParentType, astParentFullName)
+      Ast.storeInDiffGraph(Ast(typeDeclNode_), diffGraph)
+      typeDeclNode_
+    }
+
+    method.astParentFullName = parentNode.fullName
+    method.astParentType = parentNode.label
+    val functionBinding = NewBinding().name(methodName).methodFullName(methodFullName).signature(signature)
+    Ast(functionBinding).withBindsEdge(parentNode, functionBinding).withRefEdge(functionBinding, method)
+  }
   def astForFuncDecl(funcDecl: ParserNodeInfo): Seq[Ast] = {
 
     val filename = relPathFileName
     val name     = funcDecl.json(ParserKeys.Name).obj(ParserKeys.Name).str
-    val fullname = s"${fullyQualifiedPackage.get()}.${name}"
+    val fullname = s"${fullyQualifiedPackage}.${name}"
     // TODO: handle multiple return type or tuple (int, int)
     val returnType     = getReturnType(funcDecl.json(ParserKeys.Type)).headOption.getOrElse("")
     val templateParams = ""
-    val methodType     = createParserNodeInfo(funcDecl.json(ParserKeys.Type))
+    val params         = createParserNodeInfo(funcDecl.json(ParserKeys.Type)(ParserKeys.Params))
     val signature =
-      s"$fullname$templateParams (${parameterSignature(methodType)})$returnType "
+      s"$fullname$templateParams (${parameterSignature(params)})$returnType "
 
     val methodNode_ = methodNode(funcDecl, name, funcDecl.code, fullname, Some(signature), filename)
     methodAstParentStack.push(methodNode_)
     scope.pushNewScope(methodNode_)
     val astForMethod = methodAst(
       methodNode_,
-      astForMethodParameter(methodType),
+      astForMethodParameter(params),
       astForMethodBody(funcDecl.json(ParserKeys.Body)),
       newMethodReturnNode(returnType, None, line(funcDecl), column(funcDecl))
     )
     scope.popScope()
     methodAstParentStack.pop()
-    // TODO register type above
-    Seq(astForMethod)
+    val typeDeclAst = createFunctionTypeAndTypeDecl(funcDecl, methodNode_, name, fullname, signature)
+    Seq(astForMethod.merge(typeDeclAst))
   }
 
-  private def astForMethodParameter(paramType: ParserNodeInfo): Seq[Ast] = {
-    Seq(Ast(NewMethodParameterIn()))
+  private def astForMethodParameter(params: ParserNodeInfo): Seq[Ast] = {
+    var index = 1
+    params
+      .json(ParserKeys.List)
+      .arrOpt
+      .getOrElse(ArrayBuffer())
```

Comment:
`ArrayBuffer`s are mutable, are they necessary? Would prefer something immutable if possible, as the memory is assumed fixed in that case and for empty collections it would be minimal.

Reply:
Similarly in `AstForGenDeclarationCreator`

Reply:
what is the other option? I could only see upickel provind `.arr` or `arrOpt`

Reply:
I see there is no easy way to convert this to an immutable collection for free, I think we leave it for now